include/arch/common/arm_semihost.h

```diff
@@ -0,0 +1,190 @@
+ */
+
+/**
+ * @file
+ *
+ * @brief public ARM semihosting APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ARCH_COMMON_ARM_SEMIHOST_H_
+#define ZEPHYR_INCLUDE_ARCH_COMMON_ARM_SEMIHOST_H_
+
+/** @brief ARM semihosting instructions */
+enum arm_semihost_instr {
+	/*
+	 * File I/O operations
+	 */
+
+	/* Open a file or stream on the host system. */
+	ARM_SEMIHOST_OPEN       = 0x01,
+	/* Check whether a file is associated with a stream/terminal */
+	ARM_SEMIHOST_ISTTY      = 0x09,
+	/* Write to a file or stream. */
+	ARM_SEMIHOST_WRITE      = 0x05,
+	/* Read from a file at the current cursor position. */
+	ARM_SEMIHOST_READ       = 0x06,
+	/* Closes a file on the host which has been opened by ARM_SEMIHOST_OPEN. */
+	ARM_SEMIHOST_CLOSE      = 0x02,
+	/* Get the length of a file. */
+	ARM_SEMIHOST_FLEN       = 0x0C,
+	/* Set the file cursor to a given position in a file. */
+	ARM_SEMIHOST_SEEK       = 0x0A,
+	/* Get a temporary absolute file path to create a temporary file. */
+	ARM_SEMIHOST_TMPNAM     = 0x0D,
+	/* Remove a file on the host system. Possibly insecure! */
+	ARM_SEMIHOST_REMOVE     = 0x0E,
+	/* Rename a file on the host system. Possibly insecure! */
+	ARM_SEMIHOST_RENAME     = 0x0F,
+
+	/*
+	 * Terminal I/O operations
+	 */
+
+	/* Write one character to the debug terminal. */
+	ARM_SEMIHOST_WRITEC     = 0x03,
+	/* Write a 0-terminated string to the debug terminal. */
+	ARM_SEMIHOST_WRITE0     = 0x04,
+	/* Read one character from the debug terminal. */
+	ARM_SEMIHOST_READC      = 0x07,
+
+	/*
+	 * Time operations
+	 */
+	ARM_SEMIHOST_CLOCK      = 0x10,
+	ARM_SEMIHOST_ELAPSED    = 0x30,
+	ARM_SEMIHOST_TICKFREQ   = 0x31,
+	ARM_SEMIHOST_TIME       = 0x11,
+
+	/*
+	 * System/Misc. operations
+	 */
+
+	/* Retrieve the errno variable from semihosting operations. */
+	ARM_SEMIHOST_ERRNO              = 0x13,
+	/* Get commandline parameters for the application to run with */
+	ARM_SEMIHOST_GET_CMDLINE        = 0x15,
+	ARM_SEMIHOST_HEAPINFO           = 0x16,
+	ARM_SEMIHOST_ISERROR            = 0x08,
+	ARM_SEMIHOST_SYSTEM             = 0x12
+};
+
+/**
+ * @brief Modes to open a file with
+ *
+ * Behaviour corresponds to equivalent fopen strings.
+ * i.e. ARM_SEMIHOST_OPEN_RB_PLUS == "rb+"
+ */
+enum arm_semihost_open_mode {
+	ARM_SEMIHOST_OPEN_R             = 0,
+	ARM_SEMIHOST_OPEN_RB            = 1,
+	ARM_SEMIHOST_OPEN_R_PLUS        = 2,
+	ARM_SEMIHOST_OPEN_RB_PLUS       = 3,
+	ARM_SEMIHOST_OPEN_W             = 4,
+	ARM_SEMIHOST_OPEN_WB            = 5,
+	ARM_SEMIHOST_OPEN_W_PLUS        = 6,
+	ARM_SEMIHOST_OPEN_WB_PLUS       = 7,
+	ARM_SEMIHOST_OPEN_A             = 8,
+	ARM_SEMIHOST_OPEN_AB            = 9,
+	ARM_SEMIHOST_OPEN_A_PLUS        = 10,
+	ARM_SEMIHOST_OPEN_AB_PLUS       = 11,
+};
+
+/**
+ * @brief Manually execute a semihosting instruction
+ *
+ * @param instr instruction code to run
+ * @param args instruction specific arguments
+ *
+ * @return integer return code of instruction
```

Comment:
any more details what might be ok value and which is an error? Many functions bellow return directly after calling this function.

Reply:
Not really, this function is only exposed in case a user wants to run one of the other instructions. The meanings of the return codes varies by instruction, and would need to be  checked per instruction here: https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-